public_html/wp-content/plugins/wordcamp-qbo/wordcamp-qbo.php

```diff
@@ -493,9 +493,31 @@ protected static function build_qbo_create_invoice_request( int $invoice_id ) {
 		 * for the first time, so we don't need any code to automatically activate them.
 		 */
 		if ( 'USD' != $currency_code ) {
+			$response = wp_remote_get(
+				sprintf(
+					'%s/v3/company/%d/exchangerate?sourcecurrencycode=%s',
+					self::$api_base_url,
+					rawurlencode( $realm_id ),
+					$currency_code,
+				),
+				array(
+					'timeout' => self::REMOTE_REQUEST_TIMEOUT,
+					'headers' => array(
+						'Authorization' => $oauth_header,
+						'Accept'        => 'application/json',
+						'Content-Type'  => 'application/json',
+					),
+				)
+			);
+
+			$body          = json_decode( wp_remote_retrieve_body( $response ), true );
+			$exchange_rate = $body['ExchangeRate']['Rate'];
```

Comment:
what do you think about some kind of caching around the lookup too?

Reply:
> we should probably add some error handling around this - for if the API returns an error or isn't available

👍 Will add that.

> what do you think about some kind of caching around the lookup too?

看起來 Quickbook 每 4 小時就會更新一次 exchange rate，所以我們設定的話應該要更短，不然不知道他是何時開始更新的。
https://quickbooks.intuit.com/learn-support/en-ca/help-article/multicurrency/learn-multicurrency-quickbooks-online/L5krkKQi8_CA_en_CA#:~:text=Foreign%20currency%20values%20are%20automatically,Markit%2C%20an%20information%20services%20company.

It looks like QuickBooks updates the exchange rates every 4 hours, so if we were to set a cache, it needs to have a shorter interval to ensure we get the latest rates, since we don't know exactly when their updates begin.
[Source](https://quickbooks.intuit.com/learn-support/en-ca/help-article/multicurrency/learn-multicurrency-quickbooks-online/L5krkKQi8_CA_en_CA#:~:text=Foreign%20currency%20values%20are%20automatically,Markit%2C%20an%20information%20services%20company.)

The API call is triggered only when an invoice is approved. I saw that this month, the maximum number of approvals in one day is 16, and they might be in different currencies. It seems to me that we don't need to implement caching for this. What do you think?

Reply:
> we should probably add some error handling around this - for if the API returns an error or isn't available

👍 Will add that.

> what do you think about some kind of caching around the lookup too?

[It](https://quickbooks.intuit.com/learn-support/en-ca/help-article/multicurrency/learn-multicurrency-quickbooks-online/L5krkKQi8_CA_en_CA#:~:text=Foreign%20currency%20values%20are%20automatically,Markit%2C%20an%20information%20services%20company.) looks like QuickBooks updates the exchange rates every 4 hours, so if we were to set a cache, it needs to have a shorter interval to ensure we get the latest rates, since we don't know exactly when their updates begin.

The API call is triggered only when an invoice is approved. I saw that this month, the maximum number of approvals in one day is 16, and they might be in different currencies. It seems to me that we don't need to implement caching for this. What do you think?

Reply:
it was unclear how often this code was called - so not adding caching based on a low frequency makes sense. We could always add a code comment saying as much to help in the future.

Reply:
Good point.

Reply:
Comment added: 8181f66
Error handling: b500574